pkg/controller/freight/finder.go

```diff
@@ -284,3 +284,76 @@ func FindChart(
 	// from the desired origin has been promoted yet.
 	return nil, nil
 }
+
+func FindArtifact(
+	ctx context.Context,
+	cl client.Client,
+	project string,
+	freightReqs []kargoapi.FreightRequest,
+	desiredOrigin *kargoapi.FreightOrigin,
+	freight []kargoapi.FreightReference,
+	subName string,
+) (*kargoapi.ArtifactReference, error) {
+	// If no origin was explicitly identified, we need to look at all possible
+	// origins. If there's only one that could provide the artifact we're looking
+	// for, great. If there's more than one, there's ambiguity, and we need to
+	// return an error.
+	if desiredOrigin == nil {
+		for i := range freightReqs {
+			requestedFreight := freightReqs[i]
+			warehouse, err := api.GetWarehouse(
+				ctx,
+				cl,
+				types.NamespacedName{
+					Name:      requestedFreight.Origin.Name,
+					Namespace: project,
+				},
+			)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"error getting Warehouse %q in namespace %q: %w",
+					requestedFreight.Origin.Name, project, err,
+				)
+			}
+			if warehouse == nil {
+				// nolint:staticcheck
+				return nil, fmt.Errorf(
+					"Warehouse %q not found in namespace %q",
+					requestedFreight.Origin.Name, project,
+				)
+			}
+			for _, sub := range warehouse.Spec.InternalSubscriptions {
+				if sub.Subscription != nil && sub.Subscription.Name == subName {
+					if desiredOrigin != nil {
+						return nil, fmt.Errorf(
+							"multiple requested Freight could potentially provide a "+
+								"%q artifact; please update promotion steps to "+
+								"disambiguate",
+							subName,
+						)
+					}
+					desiredOrigin = &requestedFreight.Origin
```

Comment:
Nit: I don't love mutating a parameter passed in to the function as a side effect to calling the function. Probably doesn't really matter with how this function is used, but I'd prefer a local variable instead